### PR TITLE
Add MS ADPCM audio format decoding support

### PR DESCRIPTION
**Pull request by ocdi**
This PR adds support for the MS ADPCM audio format, used by the remastered music.

#18239

### CODE REVIEW

File: OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using OpenRA.Primitives;
+
+namespace OpenRA.Mods.Common.FileFormats
+{
+	public static partial class WavReader
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: Not sure about using partial classes. Ping @OpenRA/engine-hackers

File: OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	{
+		sealed class WavStreamMsAdpcm : ReadOnlyAdapterStream
+		{
+			/* format docs https://wiki.multimedia.cx/index.php/Microsoft_ADPCM
+			 */
+
+			const int MsADPCMAdaptCoeffCount = 7;
+
+			internal static int[] AdaptationTable = new[]
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: Dont think we are using internal.

File: OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+			/// Decodes a block of MS ADPCM data
+			/// </summary>
+			/// <param name="baseStream">The underlying stream to read data from</param>
+			/// <param name="samples">A block worth of PCM samples</param>
+			/// <returns>True when there is no more data or we can't continue</returns>
+			bool DecodeBlock(Stream baseStream, short[] samples)
+			{
+				int chan, k, blockindx, sampleindx;
+				short bytecode;
+				short[] bpred = new short[2], chan_idelta = new short[2];
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: `var`

File: OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs (hunk 1 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: Since this just contains a complete other class, not really a fan.

Comment (pchote) [reply OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: There doesn't appear to be any benefit to using `partial` here, so we should avoid it.

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 4]
```diff
@@ -41,6 +41,7 @@ public static bool LoadSound(Stream s, out Func<Stream> result, out short channe
 			var dataSize = -1;
 			short blockAlign = -1;
 			int uncompressedSize = -1;
+			short samplesperblock = 0;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs]: camelCase, please (`samplesPerBlock`)

File: OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+				768, 614, 512, 409, 307, 230, 230, 230
+			};
+
+			internal static int[] AdaptCoeff1 = new[] { 256, 512, 0, 192, 240, 460, 392 };
+
+			internal static int[] AdaptCoeff2 = new[] { 0, -256, 0, 64, 0, -208, -232 };
+
+			readonly short channels;
+			private readonly short blockAlign;
+			private readonly short samplesperblock;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/FileFormats/WavReader.WavStreamMsAdpcm.cs]: same here

**Comment by pchote** [comment]
This appears to be a direct translation of https://github.com/erikd/libsndfile/blob/master/src/ms_adpcm.c, which causes two problems:
* Removing the original author attribution is a bad move
* Clean C# looks a bit different to clean C, so some of the implementation details will want to change (at the zeroth order: naming conventions, but there are also patterns related to array / stream use)

I would much prefer it if we could have our own implementation that uses libsndfile as a reference for the algorithm, but is otherwise written from scratch using clean C#. This will make the code much more compatible with OpenRA's coding standards, and sidesteps the issues around license headers when copying files from other projects (but we should of course still reference libsndfile in a comment on `DecodeBlock`).

**Comment by ocdi** [comment]
Agree with your comments, will address. Was a bit eager to submit after seeing your video @pchote

**Comment by ocdi** [comment]
I've spent a bit of time to improve this and re-write this to be much more C# like, hopefully you guys agree. I ended up writing a unit test that exercises this code, comparing the original output data against my revised version, which made working on this much easier. I also tried to change this to follow the ordering and description of the linked format doc. I'm finally intimately familiar and understand what the code is doing.

One part I don't like is the many tight loops at the start of the now `BufferData` method, but the ordering is very particular and depends on whether it is mono/stereo there.

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 6]
```diff
@@ -11,14 +11,15 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:14]: Not needed.

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 7]
```diff
@@ -97,7 +101,9 @@ public static bool LoadSound(Stream s, out Func<Stream> result, out short channe
 			{
 				var audioStream = SegmentStream.CreateWithoutOwningStream(s, dataOffset, dataSize);
 				if (audioType == WaveType.ImaAdpcm)
-					return new WavStream(audioStream, dataSize, blockAlign, chan, uncompressedSize);
+					return new WavStreamImaAdpcm(audioStream, dataSize, blockAlign, chan, uncompressedSize);
+				else if (audioType == WaveType.MsAdpcm)
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:105]:
```suggestion
				if (audioType == WaveType.MsAdpcm)
```

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 8]
```diff
@@ -206,5 +212,108 @@ protected override bool BufferData(Stream baseStream, Queue<byte> data)
...
+
+				// Compute next Adaptive Scale Factor (ASF), saturating to lower bound of 16
+				idelta = (short)((AdaptationTable[nibble] * idelta) >> 8);
+				if (idelta < 16)
+					idelta = 16;
+
+				return s1;
+			}
+
+			private static short ClampInt16(int current)
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:309]: remove `private`

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 9]
```diff
@@ -206,5 +212,108 @@ protected override bool BufferData(Stream baseStream, Queue<byte> data)
...
+
+			short WriteSample(short t, Queue<byte> data)
+			{
+				data.Enqueue((byte)t);
+				data.Enqueue((byte)(t >> 8));
+				return t;
+			}
+
+			// This code contains elements from libsndfile
+			private short DecodeNibble(short nibble, byte bpred, ref short idelta, ref short s1, ref short s2)
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:290]: remove private

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 10]
```diff
@@ -206,5 +212,108 @@ protected override bool BufferData(Stream baseStream, Queue<byte> data)
...
+			{
+				this.channels = channels;
+				blockDataSize = blockAlign - channels * 7;
+				numBlocks = dataSize / blockAlign;
+			}
+
+			protected override bool BufferData(Stream baseStream, Queue<byte> data)
+			{
+				var bpred = new byte[channels];
+				var chan_idelta = new short[channels];
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:246]: Not sure about these variable names, not a big fan of using abbreviations. At least it should be named  `chanIdelta `

File: OpenRA.Mods.Common/FileFormats/WavReader.cs [hunk 11]
```diff
@@ -206,5 +212,108 @@ protected override bool BufferData(Stream baseStream, Queue<byte> data)
...
+				for (var c = 0; c < channels; c++)
+					s1[c] = baseStream.ReadInt16();
+
+				for (var c = 0; c < channels; c++)
+					s2[c] = WriteSample(baseStream.ReadInt16(), data);
+
+				for (var c = 0; c < channels; c++)
+					WriteSample(s1[c], data);
+
+				var chanr = channels > 1 ? 1 : 0;
```
Comment (teinarss) [inline OpenRA.Mods.Common/FileFormats/WavReader.cs:266]: channelNumber?

**Pull request closed without merging**